### PR TITLE
Challenge 03 Release

### PR DESCRIPTION
**Pull request by codewithsk**
- Add challenge-03/challenge-03.ipynb
- Modified release date in root README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,11 +4,11 @@ Repository containing monthly challenges in quantum computing.
 
 ## Current Challenge
 
-**Link to Problem Notebook**: [Challenge 02](https://nbviewer.jupyter.org/github/qosf/monthly-challenges/blob/main/challenge-02/qosf-monthly-challenge-02.ipynb)
+**Link to Problem Notebook**: [Challenge 03](https://github.com/qosf/monthly-challenges/blob/main/challenge-02/qosf-monthly-challenge-02.ipynb)
```
Comment (amirebrahimi) [inline README.md:7]: Needs to be updated to point to challenge-03

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 2]
```diff
@@ -0,0 +1,56 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Challenge 03: Minimum Makespan with Quantum Annealing\n",
+    "\n",
+    "Niranjan owns a catering service and receives orders one day in advance. Each of the chefs he employs is payed by the hour. In order to optimize the cost of running his business, every night, he looks at all of the dishes he needs to deliver for the next day and asks his employees to work the specific hours he requires them for.\n",
```
Comment (amirebrahimi) [inline challenge-03/qosf-monthly-challenge-03.ipynb:9]:
```suggestion
    "Niranjan owns a catering service and receives orders one day in advance. Each of the chefs he employs is paid by the hour. In order to optimize the cost of running his business, every night, he looks at all of the dishes he needs to deliver for the next day and asks his employees to work the specific hours he requires them for.\n",
```

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    "\n",
+    "Given an order queue of length $N$ with each order $i$ taking $L_i$ time to prepare, if there are $m$ chefs, design a QUBO/Ising Hamiltonian to help Niranjan find the lowest time it would take for the $m$ chefs to prepare all $N$ dishes. \n",
```
Comment (amirebrahimi) [inline challenge-03/qosf-monthly-challenge-03.ipynb:11]:
```suggestion
    "Given an order queue of length $N$ with each order $i$ taking $L_i$ time to prepare, if there are $m$ chefs, design a QUBO/Ising Hamiltonian to help Niranjan find the shortest time it would take for the $m$ chefs to prepare all $N$ dishes. \n",
```

Comment (amirebrahimi) [reply challenge-03/qosf-monthly-challenge-03.ipynb:11]: I've never seen this type of problem before, so bear with me. A few questions:
1. Are all orders the same type?
2. Does each chef have equal capabilities in cooking each meal?
3. Are there any limits on how many chefs can cook at the same time?

Can you provide sample input data and output data to aid the participant in knowing correct output?

**Review by amirebrahimi — CHANGES_REQUESTED** [review]
Thanks for preparing this. I have a requests/changes and some questions

File: challenge-03/qosf-monthly-challenge-03.ipynb (hunk 3 above)
Comment (diemilio) [reply challenge-03/qosf-monthly-challenge-03.ipynb:11]: Hi,

This is also my first time seeing this type of problem, so I also have some questions:
1. I am not familiar with QUBO at all. Is there a gentle introduction to the algorithm? I see you provided some references, but they are very lengthy and use very advanced vocabulary. Is there something we can add that guides the reader from basic understanding of quantum algorithms to a full implementation of QUBO for a specific problem?

2. Is the idea to have participants solve problem on quantum annealer? Or can a universal QC be used too? We might need to clarify this, and also provide references on how this can be done using the latter.

Comment (mstechly) [reply challenge-03/qosf-monthly-challenge-03.ipynb:11]: These are all important points – I think the answer for these answers to all of them should be "yes", as otherwise you end up with a much more complicated version of this problem.

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    "\n",
```
Comment (mstechly) [inline challenge-03/qosf-monthly-challenge-03.ipynb:12]: I think this is a little bit too abstract for many people.
Perhaps add two or three concrete cases to work on. Like: "Case 1: number of chefs: 3, tasks length: [1,2,2,2,3,4,5]". It will make it more concrete and easier for people to start working with. 
The first case should be trivial and quickly solvable by hand, so that people have a good place to start.

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    "\\[2\\] [Ising formulations of many NP problems](https://arxiv.org/pdf/1302.5843.pdf)\n",
+    "\n",
+    "\\[3\\] [D-Wave Examples](https://docs.ocean.dwavesys.com/en/stable/getting_started.html#examples)\n",
+    "\n",
+    "## References\n",
+    "\\[4\\] [Quantum Algorithms for Scientific Computing and Approximate Optimization](https://arxiv.org/pdf/1805.03265.pdf)\n",
+    "\n",
+    "\\[5\\] [From the Quantum Approximate Optimization Algorithm to a Quantum Alternating Operator Ansatz](https://arxiv.org/pdf/1709.03489.pdf)"
+   ]
```
Comment (mstechly) [inline challenge-03/qosf-monthly-challenge-03.ipynb:32]: I think a link to my blogpost about QAOA might also be helpful, as I talk about encoding problems there (self promotion FTW!)

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Challenge 03: Minimum Makespan with Quantum Annealing\n",
+    "\n",
+    "Niranjan owns a catering service and receives orders one day in advance. Each of the chefs he employs is payed by the hour. In order to optimize the cost of running his business, every night, he looks at all of the dishes he needs to deliver for the next day and asks his employees to work the specific hours he requires them for.\n",
+    "\n",
+    "Given an order queue of length $N$ with each order $i$ taking $L_i$ time to prepare, if there are $m$ chefs, design a QUBO/Ising Hamiltonian to help Niranjan find the lowest time it would take for the $m$ chefs to prepare all $N$ dishes. \n",
+    "\n",
+    "Since Niranjan wants to ensure that he can run his QUBO/Ising Hamiltonian on his QC providers' free tier for as long as possible, also minimize the number of:\n",
```
Comment (mstechly) [inline challenge-03/qosf-monthly-challenge-03.ipynb:13]: I would rephrase to:
"Since Niranjan wants to ensure that he can run his algorithms on the near term devices, keep track and try to minimize the number of:
- Gates and Qubit (for Ising Hamiltonian)
- Variables (for QUBO. For embedding on a real machine also keep track of the lenghts of chains).

**Comment by mstechly** [comment]
Some comments in the PR, some here:
1. Short intro to explain what the problem is about to those who don't know it (see below)
2. Don't want to push my own materials too hard, but this might be actually helpful, I'll let you decide: https://github.com/mstechly/quantum_tsp_tutorials
3. And you should definitely add this as a reference: https://docs.dwavesys.com/docs/latest/doc_handbook.html

Intro:

```
The core of this challenge is about translating a real-world problem into a mathematical representation that fits a quantum computer. In this challenge we encourage you to use two representations which even though equivalent, are used in different contexts. 
The first is Quadratic Unconstrained Binary Optimization (QUBO) which is used primarily for the quantum annealers (D-Wave) and the other is Ising Hamiltonian, which is often used with an algorithm called Quantum Approximate Optimization Algorithm (QAOA).

This challenge might require spending some time going through the references (especially 1-3), as we are not aware of any "quick introductions" worth mentioning. If you have hard time understanding something, feel free to reach out for help on QOSF Slack [TODO: invite link].

Unless you have a prior background with this kind of problems, we encourage you to start from QUBO and then perhaps trying Ising Hamiltonians as well, as there are more materials about it.

This is also worth mentioning that D-Wave allows you to use their machines for free for limited amount of time, so you can try running your QUBO on a real machine after signing up [here](https://cloud.dwavesys.com/leap/signup/).
```

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    "1. Gates\n",
+    "2. Qubits\n",
+    "3. Length of chains (for QUBO)\n",
+    "\n",
+    "**Optional:** \n",
+    "You can also attempt to build the Cost Hamiltonian or Mixer Hamiltonian [4,5]\n",
```
Comment (mstechly) [inline challenge-03/qosf-monthly-challenge-03.ipynb:19]: What about:
"You can also attempt to build the Cost Hamiltonian or Mixer Hamiltonian for QAOA, as described in [4, 5]."

And make some note that this is advanced.

Comment (codewithsk) [reply challenge-03/qosf-monthly-challenge-03.ipynb:19]: Prefixed with "For the advanced reader..." and made the relevant changes.

File: challenge-03/qosf-monthly-challenge-03.ipynb (hunk 5 above)
Comment (codewithsk) [reply challenge-03/qosf-monthly-challenge-03.ipynb:32]: Adding link to this blogpost: https://www.mustythoughts.com/quantum-approximate-optimization-algorithm-explained

File: README.md (hunk 1 above)
Comment (codewithsk) [reply README.md:7]: Updated. Should I add an NBViewer link instead? I wasn't sure how to do it without having the changes merged in first.

File: challenge-03/qosf-monthly-challenge-03.ipynb (hunk 3 above)
Comment (codewithsk) [reply challenge-03/qosf-monthly-challenge-03.ipynb:11]: @amirebrahimi Yes, I've added these clarifications to the questions. Thanks. 

@diemilio AFAIK, we would have to use a quantum annealer to solve these problems. Does the intro that @mstechly pointed below work?

File: challenge-03/qosf-monthly-challenge-03.ipynb (hunk 5 above)
Comment (codewithsk) [reply challenge-03/qosf-monthly-challenge-03.ipynb:32]: Added link to this blog post: https://www.mustythoughts.com/quantum-approximate-optimization-algorithm-explained

File: challenge-03/qosf-monthly-challenge-03.ipynb (hunk 7 above)
Comment (codewithsk) [reply challenge-03/qosf-monthly-challenge-03.ipynb:19]: Prefixed with: "For the advanced reader..." and made the relevant changes as well.

File: challenge-03/qosf-monthly-challenge-03.ipynb (hunk 4 above)
Comment (codewithsk) [reply challenge-03/qosf-monthly-challenge-03.ipynb:12]: Added a few example cases.

File: challenge-03/qosf-monthly-challenge-03.ipynb [hunk 8]
```diff
@@ -6,29 +6,113 @@
...
     "\n",
-    "Since Niranjan wants to ensure that he can run his QUBO/Ising Hamiltonian on his QC providers' free tier for as long as possible, also minimize the number of:\n",
-    "1. Gates\n",
-    "2. Qubits\n",
-    "3. Length of chains (for QUBO)\n",
+    "Unless you have a prior background with these kinds of problems, we encourage you to start from QUBO and then perhaps trying Ising Hamiltonians as well, as there are more materials about it.\n",
+    "\n",
+    "It is also worth mentioning that D-Wave allows you to use their machines for free for a limited amount of time, so you can try running your QUBO on a real machine after signing up [here](https://cloud.dwavesys.com/leap/signup/).\n",
+    "\n",
+    "Niranjan owns a catering service and receives orders one day in advance. Each of the chefs he employs is paid by the hour. In order to optimize the cost of running his business, every night, he looks at all of the dishes he needs to deliver for the next day and asks his employees to work the specific hours he requires them for.\n",
```
Comment (mstechly) [inline challenge-03/qosf-monthly-challenge-03.ipynb:18]: I would just add a header or sth. to make it clear that above was the "intro part" and below is "actual challenge".

**Review by mstechly — APPROVED** [review]
